sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,384 @@
+                {
+                    var totalBindings = method.ParameterList.Parameters.Count(param => param.AttributeLists.Count > 0);
+
+                    indentedTextWriter.WriteLine("var " + functionName + "Binding" + bindingCount.ToString() + " = new {");
+                    indentedTextWriter.Indent++;
+
+                    // if there are only two bindings and one is an httptrigger, then we use "$return" for the binding name
+                    if (totalBindings < 2)
+                    {
+                        indentedTextWriter.WriteLine("name = \"" + "$return" + "\",");
+                    }
+                    else
+                    {
+                        // when there are more than 2 bindings and one is an HttpTrigger, we can assume that there is another output binding
+                        // we return HttpResponse instead of "$return"
+                        indentedTextWriter.WriteLine("name = \"" + "HttpResponse" + "\",");
+                    }
+
+                    indentedTextWriter.WriteLine("type = \"" + "http" + "\",");
+                    indentedTextWriter.WriteLine("direction = \"" + "Out" + "\",");
+                    indentedTextWriter.Indent--;
+                    indentedTextWriter.WriteLine("};");
+                    indentedTextWriter.WriteLine("var " + functionName + "Binding" + bindingCount.ToString() + "JSONstring = JsonSerializer.Serialize(" + functionName + "Binding" + bindingCount.ToString() + ").ToString();");
+                    indentedTextWriter.WriteLine(functionName + "RawBindings.Add(" + functionName + "Binding" + bindingCount.ToString() + "JSONstring);");
+                    bindingCount++;
+                }
+            }
+        }
+
+        private static void AddEnumTypes(IndentedTextWriter indentedTextWriter)
+        {
+            indentedTextWriter.WriteLine("public enum AuthorizationLevel");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("Anonymous,");
+            indentedTextWriter.WriteLine("User,");
+            indentedTextWriter.WriteLine("Function,");
+            indentedTextWriter.WriteLine("System,");
+            indentedTextWriter.WriteLine("Admin");
+            indentedTextWriter.Indent--;
+            indentedTextWriter.WriteLine("}");
+        }
+
+        // adds a generated registration extension that users can call to register the source-generated function metadata provider
+        private static void AddRegistrationExtension(IndentedTextWriter indentedTextWriter)
+        {
+            indentedTextWriter.WriteLine("public static class WorkerHostBuilderFunctionMetadataProviderExtension");
+            indentedTextWriter.WriteLine("{");
+            indentedTextWriter.Indent++;
+            indentedTextWriter.WriteLine("public static IHostBuilder ConfigureGeneratedFunctionMetadataProvider(this IHostBuilder builder)");
```

Comment:
This method should be called during startup. correct? Are we planning to do it in a different PR?

Reply:
Yes it's called during startup and we're putting it in this PR because it's the registration strategy that fits our goals at the moment (opt-in, easy to call on the customer end). Without any other registration strategy the only option a customer has is to write their own DI code to add the generated type by name.